lite/proxy/validate_test.go

```diff
@@ -0,0 +1,252 @@
+					ValidatorsHash: []byte("Tendermint"),
+				},
+			},
+			commit:  lite.Commit{Header: &types.Header{Height: 11}},
+			wantErr: "Headers don't match",
+		},
+
+		{
+			meta: &types.BlockMeta{
+				Header: &types.Header{
+					Height:         11,
+					Time:           time.Date(2018, 1, 1, 1, 1, 1, 1, time.UTC),
+					ValidatorsHash: []byte("Tendermint"),
+				},
+			},
+			commit: lite.Commit{
+				Header: &types.Header{
+					Height:         11,
+					Time:           time.Date(2018, 1, 1, 1, 1, 1, 1, time.UTC),
+					ValidatorsHash: []byte("Tendermint"),
+				},
+			},
+		},
+
+		{
+			meta: &types.BlockMeta{
+				Header: &types.Header{
+					Height: 11,
+					// TODO: (@odeke-em) inquire why ValidatorsHash has to be non-blank
+					// for the Header to be hashed. Perhaps this is a security hole because
+					// an aggressor could perhaps pass in headers that don't have
+					// ValidatorsHash set and we won't be able to validate blocks.
+					ValidatorsHash: []byte("lite-test"),
+					// TODO: (@odeke-em) file an issue with Tendermint to get them to update
+					// to the latest go-wire, then no more need for this value fill to avoid
+					// the time zero value of less than 1970.
+					Time: time.Date(2018, 1, 1, 1, 1, 1, 1, time.UTC),
+				},
+			},
+			commit: lite.Commit{
+				Header: &types.Header{Height: 11, DataHash: deadBeefRipEmd160Hash},
+			},
+			wantErr: "Headers don't match",
+		},
+
+		{
+			meta: &types.BlockMeta{
+				Header: &types.Header{
+					Height: 11, DataHash: deadBeefRipEmd160Hash,
+					ValidatorsHash: []byte("Tendermint"),
+					Time:           time.Date(2017, 1, 2, 1, 1, 1, 1, time.UTC),
+				},
+			},
+			commit: lite.Commit{
+				Header: &types.Header{
+					Height: 11, DataHash: deadBeefRipEmd160Hash,
+					ValidatorsHash: []byte("Tendermint"),
+					Time:           time.Date(2017, 1, 2, 2, 1, 1, 1, time.UTC),
+				},
+				Commit: &types.Commit{BlockID: types.BlockID{Hash: []byte("DEADBEEF")}},
+			},
+			wantErr: "Headers don't match",
+		},
+
+		{
+			meta: &types.BlockMeta{
+				Header: &types.Header{
+					Height: 11, DataHash: deadBeefRipEmd160Hash,
+					ValidatorsHash: []byte("Tendermint"),
+					Time:           time.Date(2017, 1, 2, 1, 1, 1, 1, time.UTC),
+				},
+			},
+			commit: lite.Commit{
+				Header: &types.Header{
+					Height: 11, DataHash: deadBeefRipEmd160Hash,
+					ValidatorsHash: []byte("Tendermint-x"),
+					Time:           time.Date(2017, 1, 2, 1, 1, 1, 1, time.UTC),
+				},
+				Commit: &types.Commit{BlockID: types.BlockID{Hash: []byte("DEADBEEF")}},
+			},
+			wantErr: "Headers don't match",
+		},
+		// End Headers don't match test
+	}
+
+	for i, tt := range tests {
+		err := proxy.ValidateBlockMeta(tt.meta, tt.commit)
+		if tt.wantErr != "" {
+			if err == nil {
+				assert.FailNowf(t, "Unexpectedly passed", "#%d: wanted error %q", i, tt.wantErr)
+			} else {
+				assert.Contains(t, err.Error(), tt.wantErr, "#%d should contain the substring\n\n", i)
```

Comment:
Having to do substring matching on error output is brittle and not a proper integration point, this is were we would benefit from proper constructed error types that can be tests against properly.